keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -557,278 +532,231 @@ reconstructed by the driver. This has no effect on already allocated claims
+      capacity: 1
+  - name: gpu-0-third-quarter
+    attributes:
+    - name: memory
+      quantity: 10Gi
+    sharedCapacityConsumed:
+    - name: gpu-0-memory-block-2
+      capacity: 1
+  - name: gpu-0-fourth-quarter
+    attributes:
+    - name: memory
+      quantity: 10Gi
+    sharedCapacityConsumed:
+    - name: gpu-0-memory-block-3
+      capacity: 1
 ```
 
-The meaning is that the selector expression must evaluate to true for a
-particular named resource in `namedResources`.
+In this example, `gpu-0-first-half` and `gpu-0-second-half` could be allocated
+simultaneouly (because the set of `gpu-0-memory-block`s they consume are
+mutually exclusive). However, `gpu-0-first-half` and `gpu-0-first-quarter`
+could not (because `gpu-0-memory-block-0` is consumed completely by both of
+them).
 
-Future extensions could be added to support partioning of resources as well as a
-express constraints that must be satisfied *between* any selected resources. For
-example, selecting two cards which are on the same PCI root complex may be
-needed to get the required performance.
-
-Instead of defining a vendor-specific CRD, DRA driver authors (or
-administrators) could decide to allow users to create and reference
-`ResourceClaimParameters` directly within their `ResourceClaims`. This would
-avoid the translation step shown above, but at the cost of (1) providing per-
-claim configuration parameters for their requested resources, and (2) doing any
-sort of validation on the CEL expressions created by the user.
+### Using structured parameters
 
-Resource class parameters are supported the same way. To ensure that
-permissions can be limited to administrators, there's a separate cluster-scoped
-ResourceClassParameters type. Instead of individual requests, one additional
-selector can be specified there which then also must be true for all individual
-requests made with that class:
+A ResourceClaim is a request to allocate one or more devices. Each request in a
+claim may reference a pre-defined DeviceClass to narrow down which devices are
+desired or describe that directly in the request itself through one or more device
+selectors. A device selector is a CEL expression that must evaluate to true if a
+device satisfies the request. A special `device` variable provides access to
+the attributes of a device.
 
-```yaml
-kind: ResourceClassParameters
-apiVersion: resource.k8s.io/v1alpha2
+To correlate different devices, a claim may have "match attributes". Those are
+the names of attributes whose values must be the same for all devices that get
+allocated for the claim.
 
-metadata:
-  name: someArbitraryName
-
-generatedFrom:
-  name: gpu-parameters
-  kind: CardClassParameters
-  apiGroup: dra.example.com
-  uid: foobar-uid
-
-vendorParameters:
-  ...
-
-filters:
-- driverName: cards.dra.example.com
-  namedResources:
-    selector: |-
-      attributes["memory"] <= "16Gi"
-```
+Configuration parameters can be embedded in the claim so that they apply
```

Comment:
Will we have cases where only certain pods can specify certain parameters or that is out of scope?

Reply:
I see it as out-of-scope. We have separation between admin (class) and user (claim) configuration parameters and some idea how to limit access of users to those classes (not in scope for 1.31, but could come later).

Controlling what users put into claims implies a deeper understanding of them than Kubernetes has, so I see that as the role of vendor-specific admission webhook.